rest/main_test.go

```diff
@@ -32,3 +35,81 @@ func TestMain(m *testing.M) {
 
 	os.Exit(status)
 }
+
+func TestConfigOverwritesLegacyFlags(t *testing.T) {
+	osArgs := []string{
+		"sync_gateway",
+		// Legacy
+		"-verbose",
+		"-url", "1.2.3.4",
+		"-interface", "1.2.3.4",
+		// Persistent config
+		"-logging.console.log_level", "debug",
+		"-bootstrap.server", "localhost",
+		"-bootstrap.username", "test",
+
+		"config.json",
+	}
+	sc, _, _, err := parseFlags(osArgs)
+	assert.NoError(t, err)
+
+	require.NotNil(t, sc)
+	// Overwrote
+	assert.Equal(t, base.LogLevelPtr(base.LevelDebug), sc.Logging.Console.LogLevel)
+	assert.Equal(t, "localhost", sc.Bootstrap.Server)
+	// Not overwrote
+	assert.Equal(t, "1.2.3.4", sc.API.PublicInterface)
+	assert.Equal(t, "test", sc.Bootstrap.Username)
+}
+
+func TestParseFlags(t *testing.T) {
+	osArgsPrefix := []string{"sync_gateway"}
+	testCases := []struct {
+		name                            string
+		osArgs                          []string
+		expectError                     bool
+		expectedError                   error // Leave blank to not check error text
+		expectedDisablePersistentConfig *bool
+	}{
+		{
+			name:                            "Help error returned on -h",
+			osArgs:                          []string{"-h"},
+			expectError:                     true,
+			expectedError:                   flag.ErrHelp,
+			expectedDisablePersistentConfig: nil,
+		},
+		{
+			name:                            "Unknown flag",
+			osArgs:                          []string{"-unknown-flag"},
+			expectError:                     true,
+			expectedError:                   nil,
+			expectedDisablePersistentConfig: nil,
+		},
+		{
+			name:                            "Disable persistent config",
+			osArgs:                          []string{"-disable_persistent_config"},
+			expectError:                     false,
+			expectedError:                   nil,
+			expectedDisablePersistentConfig: base.BoolPtr(true),
+		},
+		{
+			name:                            "Config flag",
+			osArgs:                          []string{"-bootstrap.server", "1.2.3.4"},
+			expectError:                     false,
+			expectedError:                   nil,
+			expectedDisablePersistentConfig: base.BoolPtr(false),
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			_, _, disablePersistentConfig, err := parseFlags(append(osArgsPrefix, test.osArgs...))
+			if test.expectError {
+				assert.Error(t, err)
+				if test.expectedError != nil {
+					assert.EqualError(t, test.expectedError, err.Error())
+				}
```

Comment:
Would prefer something like this - to comprehensively cover error cases, even if that means adding an expected error for the unknown flag case.

Prior to this suggestion, the test would pass if we weren't expecting an error but actually got one.

```suggestion
			if test.expectedError != nil {
				require.Error(t, err)
				assert.Contains(t, err.Error(), test.expectedError.Error())
			} else {
				assert.NoError(t, err)
```